nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java

```diff
@@ -561,43 +548,54 @@ public static void deleteFileNowOrLater(File f) {
      * Java Runtime. Non-empty directories will be deleted recursively.
      * @param f the file to delete on JVM shutdown
      */
-    public static void deleteOnExitRecursively(File f) {
+    public static void deleteOnExit(File f) {
         if (f != null && f.exists()) {
             final SecurityManager securityManager = System.getSecurityManager();
             if (securityManager != null) {
                 securityManager.checkDelete(f.getAbsolutePath());
             }
-            FILES_TO_DELETE_ON_SHUTDOWN.add(f);
+            FILE_DELETION_ON_EXIT.add(f);
         }
     }
 
-    private static final Set<File> FILES_TO_DELETE_ON_SHUTDOWN = new LinkedHashSet<>();
+    private static final FileDeletionThread FILE_DELETION_ON_EXIT = new FileDeletionThread();
 
     static {
-        Runtime.getRuntime().addShutdownHook(
-            new Thread() {
-                @Override
-                public void run() {
-                    for (File f : FILES_TO_DELETE_ON_SHUTDOWN) {
-                        delete(f);
-                    }
-                    FILES_TO_DELETE_ON_SHUTDOWN.clear();
-                }
-                private void delete(File file) {
-                    if (file.exists()) {
-                        if (file.isDirectory()) {
-                            final File[] files = file.listFiles();
-                            if (files != null) {
-                                for (File f : files) {
-                                    delete(f);
-                                }
-                            }
-                        }
-                        file.delete();
-                    }
+        Runtime.getRuntime().addShutdownHook(FILE_DELETION_ON_EXIT);
+    }
+
+    private static final class FileDeletionThread extends Thread {
+
+        private static final Logger LOG = Logger.getLogger(FileDeletionThread.class.getName());
+        private final Set<String> filesToDelete = new LinkedHashSet<>();
+
+        public void add(File file) {
+            if (file != null && file.exists()) {
+                filesToDelete.add(file.getAbsolutePath());
+            }
+        }
+
+        @Override
+        public void run() {
+            for (String path : filesToDelete) {
+                delete(new File(path));
+            }
+            filesToDelete.clear();
+        }
+
+        private void delete(File file) {
+            if (file.exists()) {
+                try {
+                    Files.walk(file.toPath())
```

Comment:
> The stream is <i>weakly consistent</i>. It does not freeze the
>      * file tree while iterating, so it may (or may not) reflect updates to
>      * the file tree that occur after returned from this method.

Temporary directory may still exist after clean up if a new file appeared during clean up phase.

Reply:
How would you like to address this?

Reply:
I agree that every attempt may fail with possible race conditions and deleteOnExit does job as "best effort". (Windows has open descriptor to deleted file, new files appearing during hookup processing, etc.)

I don't know ho much robust have this utility be. Maybe it is ok just admit the fact in some cases there will be left some temporary directory. Or maybe not and we should spend more time with thinking about solution.

Reply:
We could still improve by having the files where deletion failed retried several times.